### PR TITLE
LibGfx/JPEG: Performance work regarding the Huffman stream

### PR DESCRIPTION
**Pull request by LucasChollet**
If this PR is too much to review in a single shot, I can easily split it!

Results of `BenchmarkJPEGLoader`:

<table>
<tr>
 <td>
 <td>small_image
 <td>big_image
 <td>rgb_image
 <td>several_scans
<tr>
 <td>Master
 <td>0.0±0.0ms
 <td>601.7±4.7ms
 <td>9.0±0.0ms
 <td>7.9±0.3ms
<tr>
 <td>Using `peek` in `read`
 <td>0.0±0.0ms
 <td>557.7±32.3ms
 <td>8.1±0.6ms
 <td>7.9±0.3ms
<tr>
 <td>Using `peek` in `next_symbol`
 <td>0.0±0.0ms
 <td>526.1±40.6ms
 <td>8.0±0.3ms
 <td>7.7±0.6ms
<tr>
 <td>Using the Lookup table
 <td>0.0±0.0ms
 <td>459.8±34ms
 <td>7.9±0.3ms
 <td>7.2±0.4ms
</table>

With this image:
https://commons.wikimedia.org/wiki/File:Seven_Coloured_Pencils_(sRGB).jpg
`image` is now only 3x (previously 4) slower than libjpeg-turbo's `djpeg`!

### CODE REVIEW

File: Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp [hunk 1]
```diff
@@ -891,6 +891,7 @@ static ErrorOr<void> read_huffman_table(Stream& stream, JPEGLoadingContext& cont
         }
 
         table.codes.ensure_capacity(total_codes);
+        table.symbols.ensure_capacity(total_codes);
```
Comment (IdanHo) [inline Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp:937]: Not a full review, just a quick nitpick: If we ensure we have enough capacity beforehand we can use an unchecked_append below to get rid of the vector resize checks.

Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp:937]: I would rather keep these checks, they don't hurt performance at all.
Actually, I did it more for readability than performance, these vectors don't contain more than a few hundreds elements.

File: Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp [hunk 2]
```diff
@@ -253,25 +253,36 @@ class HuffmanStream {
...
-        return value;
+        auto const mask = max >> (8 + m_bit_offset);
+
+        BufferType msb_buffer {};
+        if (m_byte_offset + 0 < m_stream.size())
+            msb_buffer |= ((BufferType)m_stream[m_byte_offset + 0] << (2 * 8));
+        if (m_byte_offset + 1 < m_stream.size())
+            msb_buffer |= ((BufferType)m_stream[m_byte_offset + 1] << (1 * 8));
+        if (m_byte_offset + 2 < m_stream.size())
+            msb_buffer |= ((BufferType)m_stream[m_byte_offset + 2] << (0 * 8));
```
Comment (GMTA) [inline Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp:334]: Please use C++-style casts (i.e. `static_cast<BufferType>(...)`)

File: Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp [hunk 3]
```diff
@@ -253,25 +253,36 @@ class HuffmanStream {
...
 
-            u8 const current_byte = m_stream[m_byte_offset];
-            u8 const current_bit = 1u & (current_byte >> (7 - m_bit_offset)); // MSB first.
+    u16 peek_bits(u8 count)
+    {
+        using BufferType = u32;
 
-            m_bit_offset++;
-            value = (value << 1) | current_bit;
+        constexpr static auto max = NumericLimits<BufferType>::max();
```
Comment (GMTA) [inline Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp:324]: Maybe add an include for `AK/NumericLimits.h` just to be sure

File: Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp [hunk 4]
```diff
@@ -194,7 +201,58 @@ struct HuffmanTable {
...
+        u8 size {};
+    };
+
+    ErrorOr<SymbolAndSize> symbol_from_code(u16 code) const
+    {
+        static constexpr u8 shift_for_cache = maximum_bits_per_code - bits_per_cached_code;
+
+        if (lookup_table[code >> shift_for_cache] != invalid_entry) {
+            u8 const code_length = lookup_table[code >> shift_for_cache] >> bits_per_cached_code;
+            return SymbolAndSize { (u8)lookup_table[code >> shift_for_cache], code_length };
```
Comment (GMTA) [inline Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp:219]: `static_cast<u8>(...)` here and below

File: Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp [hunk 5]
```diff
@@ -194,7 +201,58 @@ struct HuffmanTable {
...
+                u32 code_key = codes[code_offset] << (bits_per_cached_code - code_length);
+                for (u8 duplicate_count = 1 << (bits_per_cached_code - code_length); duplicate_count > 0; duplicate_count--) {
+                    lookup_table[code_key] = (code_length << bits_per_cached_code) | symbols[code_offset];
+                    code_key++;
+                }
+            }
+        }
+    }
+
+    static constexpr u16 invalid_entry = 0xFF;
```
Comment (GMTA) [inline Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp:254]: In our usual ordering, these statics come before the methods inside of this visibility scope (i.e. just after `private:`)

**Pull request merged**